traitlets/traitlets.py

```diff
@@ -525,22 +525,37 @@ def __get__(self, obj, cls=None):
             return self.get(obj, cls)
 
     def set(self, obj, value):
-        new_value = self._validate(obj, value)
+        new = self._validate(obj, value)
         try:
-            old_value = obj._trait_values[self.name]
+            old = obj._trait_values[self.name]
         except KeyError:
-            old_value = self.default_value
+            old = self.default()
 
-        obj._trait_values[self.name] = new_value
-        try:
-            silent = bool(old_value == new_value)
-        except:
-            # if there is an error in comparing, default to notify
-            silent = False
-        if silent is not True:
+        obj._trait_values[self.name] = new
+
+        change = self.change(obj, old, new)
+        
+        if change is not None:
             # we explicitly compare silent to True just in case the equality
             # comparison above returns something other than True/False
-            obj._notify_trait(self.name, old_value, new_value)
+            obj._notify_trait(change.name, change.old, change.new)
```

Comment:
I think this should be `notify_change` since `_notify_trait` just rebuilds the Change object you already have here.